code/controllers/subsystem/vis_overlays.dm

```diff
@@ -0,0 +1,78 @@
+SUBSYSTEM_DEF(vis_overlays)
+	name = "Vis contents overlays"
+	wait = 1 MINUTES
+	priority = FIRE_PRIORITY_VIS
+	init_order = INIT_ORDER_VIS
+
+	var/list/vis_overlay_cache
+	var/list/currentrun
+	var/datum/callback/rotate_cb
+
+/datum/controller/subsystem/vis_overlays/Initialize()
+	vis_overlay_cache = list()
+	rotate_cb = CALLBACK(src, .proc/rotate_vis_overlay)
+	return ..()
+
+/datum/controller/subsystem/vis_overlays/fire(resumed = FALSE)
+	if(!resumed)
+		currentrun = vis_overlay_cache.Copy()
+	var/list/current_run = currentrun
+
+	while(current_run.len)
+		var/key = current_run[current_run.len]
+		var/obj/effect/overlay/vis/overlay = current_run[key]
+		current_run.len--
+		if(overlay.unused && overlay.unused + overlay.cache_expiration < world.time)
+			qdel(overlay)
+			vis_overlay_cache -= key
+		if(MC_TICK_CHECK)
+			return
+
+//the "thing" var can be anything with vis_contents which includes images
+/datum/controller/subsystem/vis_overlays/proc/add_vis_overlay(atom/movable/thing, icon, iconstate, layer, plane, dir)
```

Comment:
I should probably move this to an atom proc but I'm unsure because this can be applied to `/image` as well